### ISSUE TITLE
Superset throws access is denied error when viewing dashboard even after signing in.

### ISSUE THREAD

**Issue by MalavikaN1** [apache/superset#28646]
### Bug description

I have a link to a dashboard. When I click on the link , it redirects to to login page. After logging in, the dashboard is displayed, but Access Denied toaster appears twice.  We have used Oauth authentication.

### How to reproduce the bug

1. Get the link to a particular dashboard.
2. Click on the link and open it in Incognito.
3. Redirect to Login, then login
4. Dashboard is displayed but access denied is shown.

### Screenshots/recordings

![Uploading Screenshot 2024-05-23 171641.png…]()

### Superset version

3.1.3

### Python version

3.9

### Node version

16

### Browser

Chrome

### Checklist

- [ ] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [ ] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [ ] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by MalavikaN1** [comment]
Any updates on this?

**Comment by MalavikaN1** [comment]
@dosu I've tried everything mentioned above, still facing the same error.

**Comment by rusackas** [comment]
The screenshot is broken, but I can't help but wonder if the user in question indeed has access to the dashboard and data. This sounds like an RBAC/config issue more than a Superset bug, I'm afraid.

**Comment by MalavikaN1** [comment]
Yes, the user has access to the dashboard and data.  I think the problem is with the initial loading. On clicking on the link since the user is not signed in, it redirects to login page, where the access denied should've been shown. Since an OAuth integration is used and the login page is different, the error message is not shown and only after signing in, the message from earlier is shown.

**Comment by rusackas** [comment]
Ahh, ok. I don't recall all the details, but the login page takes a redirection URL param, which can send the user back to the correct page after authorizing.

**Comment by MalavikaN1** [comment]
Any updates on this? How can the error message be prevented from propagating to the home page after signing in?

**Comment by MalavikaN1** [comment]
@rusackas

**Comment by rusackas** [comment]
Not sure how to repro this. I'm assuming you're using a custom security manager or a config issue of some sort. This is not an issue on Preset instances, so I don't think it's an issue with Superset itself.

**Comment by SharonCastel** [comment]
This error makes our customers wonder what is wrong here.... how can we prevent it?

**Comment by MasMadd** [comment]
Hello, I have the same problem here and I am struggling to find a solution, any ideas even to 'hide' that label?

**Comment by EinavDanielDX** [comment]
Hi, I'm experiencing a similar issue with Superset 4.0.0.

In my case, I’m using a custom security manager to implement SSO. It seems that when I try to access a dashboard before a session is created (e.g., the first time I log in for the day), Superset redirects from the dashboard to a login page and then back to the dashboard.

The issue is that the login page doesn’t exist because we use SSO, so a pop-up appears in the background even though the user logs in automatically.

![image](https://github.com/user-attachments/assets/016f8a3e-6337-450e-9089-0931de313f1f)

**Comment by nandwalritik** [comment]
Did anyone found any solution I am facing same issue while logging in as admin for the first time, although on subequent login it doesn't occurs.

**Comment by nandwalritik** [comment]
@EinavDanielDX @MalavikaN1 Did you guys found any fix, it hard to debug even becuase
* It doesn't leaves any logs.
* It occurs only first time during the day when we login, and will work fine for whole day.

@rusackas If there is any workaround please let me know.

**Comment by MalavikaN1** [comment]
@nandwalritik You can fix this by clearing out the existing client session flash messages when logging in.
Just do `session.pop('_flashes', None)`

**Comment by nandwalritik** [comment]
@MalavikaN1 Thanks for replying, not sure where should I add this. I will tell you about my implementation. Maybe you can help me after that.

We have integrated dashboard using embedded sdk, also I have overrided the login with below logic.
```    @expose('/login/', methods=['POST','GET'])
    def login(self):
            username = request.args.get("username")
            password = request.args.get("password")
            redirect_url= request.args.get("redirect")
            user = self.appbuilder.sm.auth_user_db(username, password)
            if user:
                # print("")
                login_user(user, remember=False)
                print(user,self.appbuilder.get_url_for_index)
                if redirect_url is not None:
                    return redirect(redirect_url)
                return redirect(self.appbuilder.get_url_for_index)

            elif g.user is not None and g.user.is_authenticated():
                if redirect_url is not None:
                    return redirect(redirect_url)

                return redirect(self.appbuilder.get_url_for_index)
            else:
                return super(CustomAuthDBView,self).login()
```

So Where should we do this `session.pop("_flashes",None)`?
* In the base application where we have embedded the dashboards (because we are receiving the pop ups) in that
* Or somewhere inside the login?

**Comment by rusackas** [comment]
This has gone silent for quite a while now. While I know I'm not being so helpful on this one (it's not an issue on Preset), I'm tempted to close it as not-planned since it's likely more of a config/customization issue than a core Superset bug. If anyone's facing this, we can move it to a Github Discussion, but I'm curious if it's still an issue at all still.

**Comment by EinavDanielDX** [comment]
The issue still occurs in our system, couldn't find an easy way to patch this.
If possible, please add some configuration flag/variable.

**Comment by xavier-GitHub76** [comment]
We have also the issue (with superset 4.0.2 and a custom security manager)

**Comment by SharonCastel** [comment]
We are waiting on silent so you want us to make some noise?

**Comment by sjk7524068** [comment]
> I know it is some sort of hack But I have done with `session.pop('_flashes', None)` before redirection and It worked. I have faced it when CustomSecurityManager implemented for SAML integration and Superset wants to send reports and alerts via headless login.

This is great, where should I apply this to ? --- "session.pop('_flashes', None)"

**Comment by rusackas** [comment]
Is this still an issue in v5, v6, or on `master`? The recurring _flashes workaround from @nirmal-tw suggests a leftover flash message surviving the redirect rather than a real denial, but it's hard to tell without a repro. Does this happen without a custom SecurityManager, and does the user definitely have access to the dashboard and its data? A minimal setup that reproduces it would help a lot.

**Comment by xavier-GitHub76** [comment]
I don't succed tu reproduce it with v3, v4, v5 and v6, all with custom security manager.

Details : 
I create a dataset, chart and dashboard.
I copy the url dashboard (I don't use permalink)
I use the link with incognito

**Comment by rusackas** [comment]
Every report involves a custom SecurityManager (OAuth/SAML/Keycloak) and the fix is popping leftover flask '_flashes' in the custom login. Last we heard this isn't reproducible (thanks @xavier-GitHub76) so I'll close this now.
